tests/unit/test_steps.py

```diff
@@ -424,3 +445,30 @@ def test_default_paths_not_converted_to_null():
     assert '"ResultPath": null' not in task_state.to_json()
     assert '"InputPath": null' not in task_state.to_json()
     assert '"OutputPath": null' not in task_state.to_json()
+
+
+# Test if boto3 session can fetch correct aws partition info from test environment
+def test_util_get_aws_partition():
+    aws_partition = "aws"
+    aws_cn_partition = "aws-cn"
+    default_region = None
+
+    # Boto3 used either info from ~/.aws/config or AWS_DEFAULT_REGION in environment
+    # to determine current region. We will replace/create AWS_DEFAULT_REGION with regions in
```

Comment:
couple questions here:

1. What will happen if your assertion fails after `cn-north-1` is set as the default region? will those changes persist for the next boto session because the test will bail or is there a guarantee that environment variables that are being touched will always be restored to a pristine state?
2. does this work with named profiles and other mechanisms that can be used to establish aws credentials where the `DEFAULT_REGION` is not set as an environment variable?

It's generally not preferable for unit tests to have any dependencies. Extraneous processes could be running or setting the region and this test could interfere and introduce surprising behaviour by modifying them.

i know the [`stubber`](https://botocore.amazonaws.com/v1/documentation/api/latest/reference/stubber.html#) allows you to create a session for a client with a specific region, but i'm not sure what mocking the session looks like.

Reply:
1. Good question. From my understanding this will continue to fail as the `cn-north-1` will be set as env var which will keep being effective till next run. To improve this specific issue I can save exceptions as non-fatal and assert at the very end of test so that changes got reversed before assertion.
2. Yes. I tested with both environment variable and ~/.aws/config file. Can't think of other configuration methods but please let me know if you have some in your mind.
3. I do agree the test should be on itself and should be mocked without dependency. So far I tried to mock boto3 client/session but have no luck. Will give stubber a try.